### ISSUE TITLE
Support for different line endings (Windows carriage return)

### ISSUE THREAD

**Issue by emmenlau** [swig/swig#2365]
Currently SWIG writes its generated files in text mode. Therefore the line endings are platform specific, and i.e. on Windows they will include carriage return `\r` in addition to newline `\n`. This can be a bit cumbersome when files are managed with source code revision systems like `git` that may complain about the changed line endings.

I checked the source code, and it seems that it would be quite easy to support an option that enables binary mode for output files. Basically all files are opened in `file.c` in method `DohNewFile()`. Its easy for me to patch in binary support in this place. But how would a more clean solution be implemented? Should the swig generator get a new option `--unix-line-endings`? And how would I pass that option to `DohNewFile()` so that it adds `b` to the `mode` variable?

Something like this:
```
  char binary_mode[6];

  // Append 'b' to the mode, for binary file I/O. Assume `mode` is never more than 4 characters:
  memset(binary_mode, 0, 6);
  strncpy(binary_mode, mode, 4);
  if (enable_unix_line_endings)
    binary_mode[strlen(binary_mode)] = 'b';
```

For reference, a long time ago somebody also asked this on the SWIG mailing list: https://sourceforge.net/p/swig/mailman/message/34573657/

**Comment by wsfulton** [comment]
How to support line endings has taken git quite a few iterations to come up with a good default solution. But you should be using `\r\n` on Windows and `\n` on Linux for an easy life. I can see that if you are using one Git working directory shared on both Windows and Linux platforms and using tools from both platforms, then this would cause problems. So I see some merit in providing a line ending option. 

A new option would need to support `\n`, `\r\n` and possibly `\r` too. My suggestion for the syntax for the new option is `-eol <x>, where x is cr, crlf or lf`.

**Comment by ojwb** [comment]
`\r` by itself as a line ending was only really used on classic macos, wasn't it?  I can see handling it when reading text files might still be useful in contexts where you might process files created long ago, but for generating new files it doesn't seem useful any more.

**Comment by emmenlau** [comment]
Yes, I agree to all the suggestions! Furthermore, lets not forget the current behavior for the end-of-line, which may be called `native` (`\n` on Unix, `\r\n` on Windows). If we want to also explicitly choose the current behavior, the choice would be to add a new command line option `-eol <x>`, where `x` is one of `cr`, `crlf` or `native`?

**Comment by emmenlau** [comment]
But I'm a bit at a loss how I would come from a command line option to the correct place in method `DohNewFile()`. Could you help me with this, or implement the change? The actual patch for `DohNewFile()` may be ok like specified above, but it's missing the actual command line option...

**Comment by wsfulton** [comment]
`-eol native` is fine, and native will be the default (as it is now).

Just make the file ending setting a global by adding in a static variable in DOH's file.c. Add a getter and a setter function. I can't see why a setter function is needed right now, so perhaps skip it for now. I suggest modelling it on `DohSetMaxHashExpand()` and `DohGetMaxHashExpand()`.
